hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocationServiceSupport.java

```diff
@@ -203,7 +204,11 @@ public void run() {
                 if (connection == null) {
                     continue;
                 }
-                if (connection.isHeartBeating()) {
+
+                boolean isAlive = connection.isAlive();
+                boolean isHeartbeating = connection.isHeartBeating();
+
+                if (isHeartbeating || (!isAlive && !connection.isCloseCompleted())) {
```

Comment:
Because, there is this case: heartbeating=false, live=true, for non-owner connection. If you look at heartbeatStopped method, it does not close the connection for non-owner connection. What is the correct thing to do for this case? Should we fail the outstanding invocations for such a connection? What happens if the connection temporarily goes to heartbeat failed and then recovers? 

Changed to: if (isHeartbeating || !connection.isClosed())  but we can still discuss what would be the best.